dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala

```diff
@@ -304,32 +320,85 @@ private[dynamodb] final case class DynamoDBExecutorImpl private (dynamoDb: Dynam
+                               retrievedItems      <- response.responses.mapError(_.toThrowable)
+                               _                   <- collectedItems.set(currentCollection ++ tableItemsMapToResponse(retrievedItems))
+                               _                   <- ZIO.fail(BatchRetryError()).when(responseUnprocessed.nonEmpty)
+                             } yield ())
+                             .retry(batchGetItem.retryPolicy.whileInput(thrown => thrown.isInstanceOf[BatchRetryError]))
+                             .catchSome(catchBatchRetryError)
+
+        retrievedItems  <- collectedItems.get
+        unprocessed     <- unprocessedKeys.get
+      } yield BatchGetItem.Response(responses = retrievedItems, unprocessedKeys = unprocessed)
 
   private def doPutItem(putItem: PutItem): ZIO[Any, Throwable, Unit] =
     dynamoDb.putItem(generatePutItemRequest(putItem)).unit.mapError(_.toThrowable)
 
   private def doGetItem(getItem: GetItem): ZIO[Any, Throwable, Option[Item]] =
     dynamoDb
       .getItem(generateGetItemRequest(getItem))
-      .mapError(_.toThrowable)
-      .map(_.itemValue.map(toDynamoItem))
+      .mapBoth(_.toThrowable, _.itemValue.map(toDynamoItem))
 
-  private def toDynamoItem(attrMap: ScalaMap[String, ZIOAwsAttributeValue.ReadOnly]): Item =
-    Item(attrMap.flatMap { case (k, v) => awsAttrValToAttrVal(v).map(attrVal => (k, attrVal)) })
+  private def generatePutItemRequest(putItem: PutItem): PutItemRequest =
+    PutItemRequest(
+      tableName = putItem.tableName.value,
+      item = attrMapToAwsAttrMap(putItem.item.map),
+      returnConsumedCapacity = Some(buildAwsReturnConsumedCapacity(putItem.capacity)),
+      returnItemCollectionMetrics = Some(ReturnItemCollectionMetrics.toZioAws(putItem.itemMetrics)),
+      conditionExpression = putItem.conditionExpression.map(_.toString),
+      returnValues = Some(buildAwsReturnValue(putItem.returnValues))
+    )
+
+  private def attrMapToAwsAttrMap(attrMap: ScalaMap[String, AttributeValue]): ScalaMap[String, ZIOAwsAttributeValue] =
+    attrMap.map { case (k, v) => (k, buildAwsAttributeValue(v)) }
 
-  private def generateBatchGetItemRequest(batchGetItem: BatchGetItem): BatchGetItemRequest =
+  private def generateGetItemRequest(getItem: GetItem): GetItemRequest                                               =
+    GetItemRequest(
+      tableName = getItem.tableName.value,
+      key = getItem.key.map.map { case (k, v) => (k, buildAwsAttributeValue(v)) },
+      consistentRead = Some(ConsistencyMode.toBoolean(getItem.consistency)),
+      returnConsumedCapacity = Some(buildAwsReturnConsumedCapacity(getItem.capacity)),
+      projectionExpression = toOption(getItem.projections)
+        .map(
+          _.mkString(
+            ", "
+          ) // TODO(adam): Not sure if this is the best way to combine projection expressions??? -- Feels like this should be a part of the projection expression trait?
```

Comment:
so my 50 cents on this is `GetItemRequest` is from `zio-aws` and the `projectionExpression` member is a list of `String`. I feel this list of string representation is a separate concern to the `ProjectionExpression` trait in the DSL - what do you think?

Reply:
I think it should stay outside of the `ProjectionExpression` trait. This is an old comment. I think I'll add a helper somewhere though since this same `mkString(", ")` is called in a few places and it will make future refactors easier.